lighthouse-cli/test/smokehouse/readme.md

```diff
@@ -0,0 +1,121 @@
+# Smokehouse
+
+Smokehouse is the Lighthouse end-to-end/smoke test runner. It takes in a set of URLs (usually pointing to custom-built test sites), runs Lighthouse on them, and compares the results against a set of expectations.
+
+By default this is done using the Lighthouse CLI (to exercise the full pipeline) with the tests listed in [`smokehouse/test-definitions/core-tests.js`](./test-definitions/core-tests.js).
+
+## Options
+
+See [`SmokehouseOptions`](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-cli/test/smokehouse/smokehouse.js#L23).
+
+## Test definitions
+
+| Name | Type | Description |
+| -- | -- | -- |
+| `id` | `string` | The string identifier of the test. |
+| `expectations` | `{lhr: Object, artifacts: Object}` | See below. |
+| `config` | `LH.Config.Json` (optional) | An optional Lighthouse config. If not specified, the default config is used.|
+| `runSerially` | `boolean` (optional) | An optional flag. If set to true, the test won't be run in parallel to other tests. Useful if the test is performance sensitive. |
+
+### Expectations
+
+The smoke test expectations can assert the values of the Lighthouse result (the `lhr`) and gathered `artifacts` for multiple URLs. The URL to be tested is specified in the expectations's `requestedUrl` field.
+
+The expectations are asserted as a subset of the actual results: all values in the expectectations must be in the actual results, but not all actual results must be asserted.
+
+Examples can be found in the [core tests](./test-definitions/).
+
+### Special numeric expectations
+
+If checking a number somewhere in the Lighthouse results, numeric comparisons can be used in place of a raw expected number. This allows asserting ranges or categories of numbers where the exact value isn't necessarily important, or to allow for expected variability in a test.
+
+The comparator is specified with a string, and the actual value being tested must be a number. Whitespace may be included in the string for readability.
+
+The following operators are supported:
+
+| Operator | Example|
+| -- | -- |
+| `>` | `'>0'` |
+| `>=` | `'>=5'` |
+| `<` | `'<1'` |
+| `<=` | `'<=10'` |
+| `+/-` | `'100+/-10'` |
+| `±` | `'100±10'` |
+
+### Special string expectations
+
+If checking a string somewhere in the Lighthouse results, a regular expression can be used in place of a string literal.
+
+## Pipeline
```

Comment:
there's also the wild `{0: ...}` format for asserting arrays with specific values but not caring about length - and `[...]` caring about exact length.